driver/src/contracts/stablex_auction_element.rs

```diff
@@ -0,0 +1,227 @@
+use byteorder::{BigEndian, ByteOrder};
+
+use std::collections::HashMap;
+
+use web3::types::{H160, U256};
+
+use dfusion_core::models::{BatchInformation, Order};
+
+#[derive(Debug, Default, PartialEq)]
+pub struct StableXAuctionElement {
+    valid_from: U256,
+    valid_until: U256,
+    pub sell_token_balance: u128,
+    pub order: Order,
+}
+
+impl StableXAuctionElement {
+    pub fn in_auction(&self, index: U256) -> bool {
+        self.valid_from <= index && index <= self.valid_until
+    }
+
+    pub fn from_bytes(order_count: &mut HashMap<H160, u16>, bytes: &[u8; 113]) -> Self {
+        let account_id = H160::from(&bytes[0..20]);
+
+        // these go together (since sell_token_balance is emitted as u256 and treated as u128
+        let sell_token_balance = BigEndian::read_u128(&bytes[36..52]);
+        let sell_token_balance_padding = BigEndian::read_u128(&bytes[20..36]);
+        assert_eq!(
+            sell_token_balance_padding, 0,
+            "User has too large balance to handle."
+        );
+
+        let buy_token = u16::from_le_bytes([bytes[53], bytes[52]]);
+        let sell_token = u16::from_le_bytes([bytes[55], bytes[54]]);
+        let valid_from = U256::from(u32::from_le_bytes([
+            bytes[59], bytes[58], bytes[57], bytes[56],
+        ]));
+        let valid_until = U256::from(u32::from_le_bytes([
+            bytes[63], bytes[62], bytes[61], bytes[60],
+        ]));
+        let is_sell_order = bytes[64] > 0;
+        let numerator = BigEndian::read_u128(&bytes[65..81]);
+        let denominator = BigEndian::read_u128(&bytes[81..97]);
+        let amount = BigEndian::read_u128(&bytes[97..113]);
+        let (buy_amount, sell_amount) =
+            compute_buy_sell_amounts(numerator, denominator, amount, is_sell_order);
+        let order_counter = order_count.entry(account_id).or_insert(0);
+        *order_counter += 1;
+        StableXAuctionElement {
+            valid_from,
+            valid_until,
+            sell_token_balance,
+            order: Order {
+                batch_information: Some(BatchInformation {
+                    slot_index: *order_counter - 1,
+                    slot: U256::from(0),
+                }),
+                account_id,
+                buy_token,
+                sell_token,
+                buy_amount,
+                sell_amount,
+            },
+        }
+    }
+}
+
+fn compute_buy_sell_amounts(
+    numerator: u128,
+    denominator: u128,
+    amount: u128,
+    is_sell_order: bool,
+) -> (u128, u128) {
+    if denominator > 0 {
+        let other = (numerator * amount) / denominator;
```

Comment:
This overflows, please fix before merging.

Reply:
Good catch, is there a rusty way to prevent overflow?

Reply:
Furthermore, even without overflow, this formula is not correct.

Notice, from the  previous version of the code, that other is computed differently in different scenarios.

```rs
        let mut other = 0;
        let (buy_amount, sell_amount) = if is_sell_order {
            if denominator > 0 {
                other = (numerator * remaining) / denominator;
            }
            (remaining, other)
        } else {
            if numerator > 0 {
                other = (denominator * remaining) / numerator;
            }
            (other, remaining)
        };
```

Reply:
Yeah,
I guess it depends on how we gonna store the price for a buyOrder, do we us /pi or 1/pi. I guess your definition is closer to the one Tom was using in his papers.

I think we would not implement any of this logic in rust, as long as we do not have any in the smart contract. 

Probably, I should have taken isSellOrder completely out of the interface of smartcontract ...

Reply:
That would be a cool fun thing to do! Although the semantic of buy and sell orders are completely different and not symmetric (since sell orders are unbounded and buy orders are bounded). This means we can't treat them in the same way.